### PR TITLE
Upgrade config, UI, and command package patterns

### PR DESCRIPTION
**Pull request by wagoodman**
Syft is several months ahead regarding patterns for configuration management and `cmd` package needs. This PR brings the differences closer to 0.

Specifically:
- Splits up the configuration into multiple files and adds the `parseConfigValues ` & `parseConfigValues` methods on each config struct that needs them.
- Adds the event loop that separates common event handling (signals, worker errors, and bus events) concerns from the UI implementations.
- Changes the main cobra execution path to prefer raising errors (use `RunE` over `Run`).
- Reorganize `ui` package to a more flat structure

### CODE REVIEW

File: internal/config/application.go [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+			switch v := cfg.CliOptions.Verbosity; {
+			case v == 1:
+				cfg.Log.LevelOpt = logrus.InfoLevel
+			case v >= 2:
+				cfg.Log.LevelOpt = logrus.DebugLevel
+			default:
+				cfg.Log.LevelOpt = logrus.ErrorLevel
+			}
+		}
+	}
```
Comment (luhring) [inline internal/config/application.go:129]: nit: this block could be factored out as a function that performs a single task: determining the log level

File: internal/config/registry.go [hunk 2]
```diff
@@ -22,7 +24,14 @@ type registry struct {
...
 
-func (cfg *registry) parseConfigValues() {
+func (cfg registry) loadDefaultValues(v *viper.Viper) {
+	v.SetDefault("registry.insecure-skip-tls-verify", false)
+	v.SetDefault("registry.insecure-use-http", false)
+	v.SetDefault("registry.auth", []RegistryCredentials{})
+}
+
+// nolint: unparam
+func (cfg *registry) parseConfigValues() error {
```
Comment (luhring) [inline internal/config/registry.go:34]: question: I looked up `unparam` on the [golangci-lint docs](https://golangci-lint.run/usage/linters/) — could it be we no longer need this? The descrption of `unparam` is:

>Reports unused function parameters

File: test/cli/json_descriptor_test.go [hunk 3]
```diff
@@ -19,9 +19,9 @@ func TestJsonDescriptor(t *testing.T) {
 				"GRYPE_CHECK_FOR_APP_UPDATE": "false",
 			},
 			assertions: []traitAssertion{
-				assertInOutput(`"CheckForAppUpdate": false`), // app config
-				assertInOutput(`"db": {`),                    // db status
-				assertInOutput(`"built":`),                   // db status
+				assertInOutput(`"check-for-app-update": false`), // app config
+				assertInOutput(`"db": {`),                       // db status
+				assertInOutput(`"built":`),                      // db status
```
Comment (luhring) [inline test/cli/json_descriptor_test.go:24]: question: What do the comments (both `db status`) on these two lines mean?

**Review by luhring — APPROVED** [review]
Looks good! Thanks for tackling this. I left a couple of comments and questions for discussion, but none of these block the merge itself.

File: test/cli/json_descriptor_test.go (hunk 3 above)
Comment (wagoodman) [reply test/cli/json_descriptor_test.go:24]: I'll add more verbiage here. Essentially this is saying that these assertions are for the `db` block in the json output.

**Comment by wagoodman** [comment]
In making the above updates I noticed that I missed adding struct tags in some config options --I've updated https://github.com/anchore/grype/pull/406/commits/999aaa78ca3a98116d082bc1fb27bfd5642ad43d

File: internal/config/application.go (hunk 1 above)
Comment (wagoodman) [reply internal/config/application.go:159]: I broke out the remaining option parsing blocks into functions

File: internal/config/registry.go (hunk 2 above)
Comment (wagoodman) [reply internal/config/registry.go]: actually it is still required such that this fulfills the `parser` interface (unparam applies to return values as well, in this case is always nil)

File: internal/config/application.go [hunk 4]
```diff
@@ -91,13 +91,43 @@ func (cfg Application) loadDefaultValues(v *viper.Viper) {
...
+	// parse application config options
+	for _, optionFn := range []func() error{
+		cfg.parseScopeOption,
+		cfg.parseLogLevelOption,
+		cfg.parseFailOnOption,
+	} {
+		if err := optionFn(); err != nil {
+			return err
+		}
+	}
```
Comment (luhring) [inline internal/config/application.go:103]: Nice idea! 👍

File: internal/config/database.go [hunk 5]
```diff
@@ -10,10 +10,10 @@ import (
...
 
 type database struct {
-	Dir                   string `mapstructure:"cache-dir"`
-	UpdateURL             string `mapstructure:"update-url"`
-	AutoUpdate            bool   `mapstructure:"auto-update"`
-	ValidateByHashOnStart bool   `mapstructure:"validate-by-hash-on-start"`
+	Dir                   string `yaml:"cache-dir" json:"cache-dir" mapstructure:"cache-dir"`
+	UpdateURL             string `yaml:"update-url" json:"update-url" mapstructure:"update-url"`
+	AutoUpdate            bool   `yaml:"auto-update" json:"auto-update" mapstructure:"auto-update"`
+	ValidateByHashOnStart bool   `yaml:"validate-by-hash-on-start" json:"validate-by-hash-on-start" mapstructure:"validate-by-hash-on-start"`
```
Comment (luhring) [inline internal/config/database.go:16]: Blocking: for JSON, we use `camelCase`

File: test/cli/json_descriptor_test.go [hunk 6]
```diff
@@ -19,9 +19,9 @@ func TestJsonDescriptor(t *testing.T) {
 				"GRYPE_CHECK_FOR_APP_UPDATE": "false",
 			},
 			assertions: []traitAssertion{
-				assertInOutput(`"check-for-app-update": false`), // app config
-				assertInOutput(`"db": {`),                       // db status
-				assertInOutput(`"built":`),                      // db status
+				assertInOutput(`"check-for-app-update": false`), // assert existence of the app config block
+				assertInOutput(`"db": {`),                       // assert existence of the db status block
+				assertInOutput(`"built":`),                      // assert existence of the db status block
```
Comment (luhring) [inline test/cli/json_descriptor_test.go:24]: Very helpful, thanks!

**Review by luhring — CHANGES_REQUESTED** [review]

File: internal/config/database.go (hunk 5 above)
Comment (luhring) [reply internal/config/database.go:16]: Disregard, and I'll approve again: I see that we tend to use `snake-case` for our configuration JSON.

So given that, I'd suggest that as we think about important but breaking changes we want to make for 1.0, we consider using a consistent JSON convention between our inputs and outputs. 🙏

**Review by luhring — APPROVED** [review]

File: internal/config/database.go (hunk 5 above)
Comment (wagoodman) [reply internal/config/database.go:16]: that makes sense 💯

**Pull request merged**